README.md
# Record-parsing and view demo

This is a demo of record reading and parsing. 

## Requirements

The app is currently uncompiled, and will require Leiningen to run. See https://github.com/technomancy/leiningen

## Installation

There is nothing to install. As this is an example for demonstration purposes, most of this is intended to be run from the Clojure REPL. However, a basic view can be created from the command line app. Give the app the path of any files you would like to view, with the full path in quotes.

### Example
Using Leiningen, the app can be run with the `lein run` command, then the file paths in quote marks, all separated by spaces, like so:

`lein run "./resources/comma-data.csv" "./resources/space-data.txt" "./resources/pipe-data.txt"`

The app should present the records from those three files sorted in reverse-email order.

## File Formats

The app should parse data from comma-separated lists, pipe-separated lists, or simple data parsed with spaces.

## Testing

Testing can be shown through Leiningen with `lein test`.

## Author

This has been created by Scott W. Starkey. For more information, contact me at yekrats@gmail.com

## License

Copyright © 2021

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-04-03
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-04-03
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/homework/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/homework/compare/0.1.0...0.1.1


test/homework/core_test.clj
(ns homework.core-test
  (:require [clojure.test :refer :all]
            [homework.core :refer :all]))

(deftest basic-parse
  (testing
    (is (= (select-keys (parse-line "Starkey | Scott | name@gmail.com | green | 01/01/2000") [:last-name :first-name :email :favorite-color :date-of-birth]) ; pipe-delimited data
           {:last-name "Starkey", :first-name "Scott", :email "name@gmail.com", :favorite-color "green", :date-of-birth "2000/01/01"}))

    (is (= (select-keys (parse-line "Starkey, Scott, name@gmail.com, green, 01/01/2000") [:last-name :first-name :email :favorite-color :date-of-birth]) ; comma-delimited data
           {:last-name "Starkey", :first-name "Scott", :email "name@gmail.com", :favorite-color "green", :date-of-birth "2000/01/01"}))

    (is (= (select-keys  (parse-line "Starkey Scott name@gmail.com green 01/01/2000") [:last-name :first-name :email :favorite-color :date-of-birth]) ; space-delimited data
           {:last-name "Starkey", :first-name "Scott", :email "name@gmail.com", :favorite-color "green", :date-of-birth "2000/01/01"}))
    ))

 (deftest file-parse
   (testing
       (is [= '({:last-name "Starkey",
                  :first-name "Scotticus",
                  :email "yekrats@gmail.com",
                  :favorite-color "green",
                  :date-of-birth "2000/01/01"}
                {:last-name "McClane",
                  :first-name "John",
                  :email "jmcclane@nypd.com",
                  :favorite-color "brown",
                  :date-of-birth "1960/12/24"}
                {:last-name "Gennero",
                  :first-name "Holly",
                  :email "gennero@nakatomi.com",
                  :favorite-color "yellow",
                  :date-of-birth "1960/06/20"}
                {:last-name "Gruber",
                  :first-name "Hans",
                  :email "hans@volksfrei.de",
                  :favorite-color "blue",
                  :date-of-birth "1958/02/14"})
            (map #(select-keys % [:last-name :first-name :email :favorite-color :date-of-birth])
                 (parse-files "./resources/comma-data.csv"))])))

(deftest date-asc
  (testing
      (is [= '({:first-name "Hans", :date-of-birth "1958/02/14"}
               {:first-name "Holly", :date-of-birth "1960/06/20"}
               {:first-name "John", :date-of-birth "1960/12/24"}
               {:first-name "Scotticus", :date-of-birth "2000/01/01"})
           (map #(select-keys % [:first-name :date-of-birth])
                (sort-by-birthdate-asc (parse-files "./resources/comma-data.csv")))])))

(deftest email-desc
  (testing
      (is [= '({:last-name "Stark", :email "tony@stark.com"}
               {:last-name "AStarkey", :email "scotto@yekrats.com"}
               {:last-name "BStarkey", :email "scotto@yekrats.com"}
               {:last-name "CStarkey", :email "scotto@yekrats.com"}
               {:last-name "Banner", :email "bruce_b@avengers.org"})
           (map #(select-keys % [:last-name :email])
                (sort-by-email-desc (parse-files "./resources/weird-space-same-email-case.txt")))
           ])))

(deftest parse-multiple-files
  (testing
      (is [= '({:last-name "Starkey"}
               {:last-name "McClane"}
               {:last-name "Gennero"}
               {:last-name "Gruber"}
               {:last-name "Starkey"}
               {:last-name "Jordan"}
               {:last-name "Bunny"}
               {:last-name "Swackhammer"})
           (map #(select-keys % [:last-name])
                (parse-files "./resources/comma-data.csv" "./resources/space-data.txt"))])))

(def testdata '({:last-name "Gruber",
  :first-name "Hans",
  :email "hans@volksfrei.de",
  :favorite-color "blue",
  :date-of-birth "1958/02/14",
  :id "05b6169f-a732-4ced-9e27-bf02a9ccc7f5"}
 {:last-name "Gennero",
  :first-name "Holly",
  :email "gennero@nakatomi.com",
  :favorite-color "yellow",
  :date-of-birth "1960/06/20",
  :id "93ef22cf-8d62-4b6c-b446-fb9cdd507985"}))

(deftest basic-views
  (testing (is (nil? (view-by-last-name-asc testdata))
               "View-by-last-name-asc should return nil and show a list of entries."))
  (testing (is (nil? (view-by-email-desc testdata))
               "View-by-email-desc should return nil and show a list of entries."))
  (testing (is (nil? (view-by-birthdate-asc testdata))
               "View-by-last-name-asc should return nil and show a list of entries."))
  (testing (is (nil? (view-by-email-desc (parse-files "./resources/comma-data.csv" "./resources/pipe-data.txt"))))
               "Should show multiple files parsed in reverse email order.")
  (testing (is (nil? (view-by-email-desc (parse-files "./resources/weird-space-same-email-case.txt"))))
               "Should show collision of email addresses sorted by last name."))

(deftest handler-test
  (testing (is [= (post-records-handler nil "93ef22cf-8d62-4b6c-b446-fb9cdd507985")
                 '{:status 200, :headers {}, :body
                    ({:last-name "Gennero",
                    :first-name "Holly",
                    :email "gennero@nakatomi.com",
                    :favorite-color "yellow",
                    :date-of-birth "1960/06/20",
                    :id "93ef22cf-8d62-4b6c-b446-fb9cdd507985"})}
                ] "Should just display one record.")))


src/homework/core.clj
(ns homework.core
  (:gen-class)
  (:require [clojure.string :as s]
            [ring.middleware.json :refer [wrap-json-response]]
            [ring.util.response :refer [response]]
            [compojure.core :refer [defroutes GET POST]]))

(defn uuid [& _] (.toString (java.util.UUID/randomUUID)))

(defn parse-date-to-ymd
  "Formats string date M/D/YYYY format to YYYY/MM/DD format (for sorting)"
  [string-date]
  (as-> (s/split string-date #"/") $
       (map #(Integer/parseInt %) $)
       (map #(format "%02d" %) $)
       [(last $) (first $) (second $)]
       (s/join "/" $)))

(defn parse-date-to-mdy
  "Formates string date YYYY/MM/DD to M/D/YYYY format (for viewing)"
  [string-date]
  (as-> (s/split string-date #"/") $
      (map #(Integer/parseInt %) $)
      [(second $) (last $) (first $) ]
      (s/join "/" $))
  )

(defn parse-line
  "Basic parsing of a line (string) into initial map."
  [string]
  (as-> (s/split string #"[\|\s\,]+") $ ;; parse basic string splitting by the pipe, space, or comma characters
      (zipmap [:last-name :first-name :email :favorite-color :date-of-birth] $) ; combine fields with data
      (update-in $ [:date-of-birth] parse-date-to-ymd)
      (update-in $ [:id] uuid)))

(defn parse-files
  "Parsing each of the lines of multiple files. Each 'file-name' is considered
  a string. Returns a map of each line processed through parse-line."
  ; Example:  (parse-files "./resources/comma-data.csv")
  [file-name & more-files]
  {:pre [(= true (or (nil? file-name) (string? file-name)))]}
  (concat
    (when file-name
      (->> (slurp file-name)
        (s/split-lines)
        (map parse-line)))
    (when more-files
      (apply parse-files more-files))))

(defn sort-by-email-desc
  "Sorts a record map by email order descending (then by last name ascending).
   Returns new map in sorted order."
  [record-map]
  (sort-by (juxt :email :last-name)
     (fn [[x0 x1] [y0 y1]]
       (let [c (- (compare x0 y0))]
         (if-not (zero? c)
           c
           (compare x1 y1))))
      record-map))

(defn generic-view
  "Step through the records with the 5 points of vital information,
  and print them out to the screen."
  [record-map]
  (as-> record-map $
    (map #(update-in % [:date-of-birth] parse-date-to-mdy) $)
    (map #(select-keys % [:last-name :first-name :email :favorite-color :date-of-birth]) $)
    (doseq [rec $] (println (str "Name: " (:first-name rec) " " (:last-name rec)
                                 ", Email: " (:email rec)
                                 ", Favorite color: " (:favorite-color rec)
                                 ", Date of birth: " (:date-of-birth rec))))))

(defn view-by-email-desc
  "Print record map in email-descending order of the 5 chief keys by means of the generic-view."
  [record-map]
  (-> record-map sort-by-email-desc generic-view))

(defn sort-by-birthdate-asc
  "Sorts a record map by email order descending (then by last name ascending).
   Returns new map in sorted order."
  [record-map]
  (sort-by :date-of-birth record-map)
  )

(defn view-by-birthdate-asc
  "Print record map of birthdate order of the 5 chief keys through the generic view."
  [record-map]
  (-> record-map sort-by-birthdate-asc generic-view))

(defn sort-by-last-name-asc
  "Sorts a record map by email order descending (then by last name ascending).
   Returns new map in sorted order."
  [record-map]
  (sort-by :last-name record-map))

(defn view-by-last-name-asc
  "Print record map of last-name order of the 5 chief keys through the generic view."
  [record-map]
  (-> record-map sort-by-last-name-asc generic-view))

(defn handler [request]
  (response request))

(def json-wrapper
  (wrap-json-response handler))

(def testdata2 '({:last-name "Gruber",
  :first-name "Hans",
  :email "hans@volksfrei.de",
  :favorite-color "blue",
  :date-of-birth "1958/02/14",
  :id "05b6169f-a732-4ced-9e27-bf02a9ccc7f5"}
 {:last-name "Gennero",
  :first-name "Holly",
  :email "gennero@nakatomi.com",
  :favorite-color "yellow",
  :date-of-birth "1960/06/20",
  :id "93ef22cf-8d62-4b6c-b446-fb9cdd507985"}))

(defn email-view-handler [request]
  (response (sort-by-email-desc testdata2)))

(defn birthdate-view-handler [request]
  (response  (sort-by-birthdate-asc testdata2)))

(defn lastname-view-handler [request]
   (response  (sort-by-last-name-asc testdata2)))

(defn post-records-handler [request id]
  (response (filter #(= id (:id %)) testdata2)))

(defn -main
  "Pulls in files from paths, and exports a view in email-ascending order."
  [& args]
  (-> (apply parse-files args)
      view-by-email-desc))
